modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,475 @@
+
+            GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(reqId);
+
+            if (clusterOpFut != null) {
+                Map<ClusterNode, Exception> errors0 = collectErrors(errors, ctx != null ? ctx.req.nodes() : null);
+
+                finishClusterFutureWithErr(clusterOpFut, err, errors0);
+            }
+
+            return;
+        }
+
+        if (ctx == null || !ctx.req.operationalNodeId().equals(kctx.localNodeId()))
+            return;
+
+        Map<ClusterNode, List<SnapshotMetadata>> metas = new HashMap<>();
+
+        try {
+            results.forEach((nodeId, nodeRes) -> {
+                // A node might be non-baseline (not required).
+                if (ctx.req.nodes().contains(nodeId)) {
+                    assert nodeRes != null && nodeRes.partsHashes == null;
+                    assert kctx.cluster().get().node(nodeId) != null;
+
+                    metas.put(kctx.cluster().get().node(nodeId), nodeRes.metas);
+                }
+            });
+
+            SnapshotMetadataVerificationTaskResult metasRes = new SnapshotMetadataVerificationTaskResult(
+                metas,
+                SnapshotChecker.reduceMetasResults(ctx.req.snapshotName(), ctx.req.snapshotPath(), metas, null,
+                    kctx.cluster().get().localNode().consistentId())
+            );
+
+            if (!F.isEmpty(metasRes.exceptions()))
+                throw new IgniteSnapshotVerifyException(metasRes.exceptions());
+        }
+        catch (Throwable th) {
+            err = th;
+        }
+
+        phase2PartsHashes.start(reqId, new SnapshotCheckProcessRequest(ctx.req, err, metas));
+
+        if (log.isDebugEnabled())
+            log.debug("Started partitions validation as part of the snapshot checking [req=" + ctx.req + ']');
+    }
+
+    /** Finds current snapshot name from the metas. */
+    private @Nullable String snpName(@Nullable Map<UUID, CheckResultDTO> results) {
+        if (F.isEmpty(results))
+            return null;
+
+        for (CheckResultDTO nodeRes : results.values()) {
+            if (nodeRes == null || F.isEmpty(nodeRes.metas))
+                continue;
+
+            assert nodeRes.metas.get(0) != null : "Empty snapshot metadata in the results";
+            assert !F.isEmpty(nodeRes.metas.get(0).snapshotName()) : "Empty snapshot name in a snapshot metadata.";
+
+            return nodeRes.metas.get(0).snapshotName();
+        }
+
+        return null;
+    }
+
+    /** Starts the snapshot full validation. */
+    public IgniteInternalFuture<SnapshotPartitionsVerifyTaskResult> start(
+        String snpName,
+        @Nullable String snpPath,
+        @Nullable Collection<String> grpNames,
+        boolean inclCstHndlrs
+    ) {
+        assert !F.isEmpty(snpName);
+
+        UUID reqId = UUID.randomUUID();
+
+        List<UUID> requiredNodes = new ArrayList<>(F.viewReadOnly(kctx.discovery().discoCache().aliveBaselineNodes(), F.node2id()));
```

Comment:
Why not Set?

Reply:
> Why not Set?

AbstractSnapshotOperationRequest creates set: `this.nodes = new HashSet<>(nodes);`

Reply:
You add an element to the `requiredNodes` which is likely will be duplicated. It's quite weird to see `List`, especially when `SnapshotCheckProcessRequest` requires `Collection`. Let's explicitly use `Set`.